### PR TITLE
Add support for latest Apache Beam SDK in Dataflow operators

### LINKED ISSUE #8630: Apache Beam 2.20.0 will break the DataFlowOperator
**Apache Airflow version**: 1.10.10

**Kubernetes version (if you are using kubernetes)** (use `kubectl version`):

**Environment**:

- **Cloud provider or hardware configuration**: GCP
- **OS** (e.g. from /etc/os-release): Ubuntu
- **Kernel** (e.g. `uname -a`): 5.0.0-1034-gcp
- **Install tools**: apache-beam (Python) on Google Cloud DataFlow
- **Others**:

**What happened**: When using Apache Beam with Python and upgrading to the latest `apache-beam=2.20.0` version, the DataFlowOperatow will always yield a failed state. The job itself runs fine. The reason is that the Monitoring String Format changed, making the dataflow hook fail retrieving the job_id from GCP.

**What you expected to happen**: The dataflow job to terminate with no error

 The job_id is not retrieved properly since the format changed

**How to reproduce it**: Set up airflow with Dataflow and install apache-beam 2.20.0 for Python. Run a DataFlow job.

**Anything else we need to know**: The commit that changes the format can be found here: https://github.com/apache/beam/commit/d38e87b6180cb60752be24bdbbb4e17daf0fb3aa
The current regexp in Airflow is `br'.*console.cloud.google.com/dataflow.*/jobs/([a-z|0-9|A-Z|\-|\_]+).*')` and can be found here for 1.10.10: https://github.com/apache/airflow/blob/v1-10-stable/airflow/contrib/hooks/gcp_dataflow_hook.py#L145
As now the `region_id` is coming after `jobs/`, we retrieve the `region_id`.
I got it to work again by using the following regexp: `br".*console.cloud.google.com/dataflow.*/jobs/.*/([a-z|0-9|A-Z|\-|\_]+).*"`

### PR DESCRIPTION
**Pull request by mik-laj**
Fix: https://github.com/apache/airflow/issues/8630

---
Make sure to mark the boxes below before creating PR: [x]

- [X] Description above provides context of the change
- [X] Unit tests coverage for changes (not needed for documentation changes)
- [X] Target Github ISSUE in description if exists
- [X] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [X] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

File: tests/providers/google/cloud/hooks/test_dataflow.py [hunk 1]
```diff
@@ -1026,14 +1027,83 @@ def test_dataflow_job_cancel_job(self):
...
-            'jobs/{}?project=XXX'.format(TEST_JOB_ID)
-        ]
+    @parameterized.expand([
+        (APACHE_BEAM_V_2_14_0_JAVA_SDK_LOG, ),
+        (APACHE_BEAM_V_2_22_0_JAVA_SDK_LOG, ),
+        (APACHE_BEAM_V_2_14_0_PYTHON_SDK_LOG, ),
+        (APACHE_BEAM_V_2_22_0_PYTHON_SDK_LOG, ),
+    ], name_func=lambda func, num, p: f"{func.__name__}_{num}")
+    def test_data_flow_valid_job_id(self, log):
+        echos = ";".join([f"echo {shlex.quote(line)}" for line in log.split("\n")])
```
Comment (mik-laj) [inline tests/providers/google/cloud/hooks/test_dataflow.py:1105]: When developing, I introduce such a change to simulate situations more similar to the real world. 
```suggestion
        echos = ";sleep 10;".join([f"echo {shlex.quote(line)}" for line in log.split("\n")])
```

File: airflow/providers/google/cloud/hooks/dataflow.py [hunk 2]
```diff
@@ -319,48 +319,50 @@ def __init__(
...
-            if line:
-                self.log.warning(line[:-1])
-            return line
-
-        if fd == self._proc.stdout.fileno():
-            line = self._proc.stdout.readline().decode()
-            if line:
-                self.log.info(line[:-1])
-            return line
+    def process_fb(self, fd):
```
Comment (turbaszek) [inline airflow/providers/google/cloud/hooks/dataflow.py]:
```suggestion
    def process_fd(self, fd):
```
Typo? Btw What is the `fd`?

Comment (mik-laj) [reply airflow/providers/google/cloud/hooks/dataflow.py]: File descriptior

> A file descriptor is either a socket or file object, or a small integer
    gotten from a fileno() method call on one of those.

File: airflow/providers/google/cloud/hooks/dataflow.py [hunk 3]
```diff
@@ -369,35 +371,31 @@ def wait_for_done(self) -> Optional[str]:
...
-            if process_ends:
-                break
             if self._proc.poll() is not None:
-                # Mark process completion but allows its outputs to be consumed.
-                process_ends = True
+                break
+
+        # Corner case: check if more output was created between the last read and the process termination
+        for readable_fb in reads:
+            self.process_fb(readable_fb)
```
Comment (turbaszek) [inline airflow/providers/google/cloud/hooks/dataflow.py]:
```suggestion
        for readable_fd in reads:
            self.process_fd(readable_fd)
```
Right?

File: airflow/providers/google/cloud/hooks/dataflow.py (hunk 2 above)
Comment (mik-laj) [reply airflow/providers/google/cloud/hooks/dataflow.py]: I upgrade the PR.

File: airflow/providers/google/cloud/hooks/dataflow.py (hunk 3 above)
Comment (mik-laj) [reply airflow/providers/google/cloud/hooks/dataflow.py]: I replaced all "fb" to "fd".

**Review by turbaszek — APPROVED** [review]

**Pull request merged**